### PR TITLE
intrepid2: allow execution space to be passed to basis getValues

### PR DESCRIPTION
**Pull request by romintomasetti**
@trilinos/intrepid2

## Motivation

More and more routines are executed in their own execution space instance. This is not yet the case for *e.g.* `Intrepid2::Basis::getValues`.

This PR fixes this.

Note that in our use case, we are running some routines in a Cuda stream that is not the default one, but in this routine we call `Intrepid2::Basis::getValues` that runs on the default stream for now. Right after calling `Intrepid2::Basis::getValues`, we use it in a kernel that is launched on the non-default stream. Clearly, our code needs a `execution_space{}.fence()` before we launch the kernel, to ensure that kernels launched on the default stream by `Intrepid2::Basis::getValues` are finished.

## Testing

I guess it should not break anything, since I kept the default signature of `Intrepid2::Basis::getValues` that calls the variant that takes a space instance with `execution_space{}` (*i.e.* the default stream).

### CODE REVIEW

**Comment by romintomasetti** [comment]
@mperego Would you have time to go over this proposition? I haven't covered all the `Intrepid2` bases yet, because I first wanted to ensure the way I make the changes is satisfying for you :wink:

**Comment by romintomasetti** [comment]
@maartenarnst will be interested by this one.

**Comment by mperego** [comment]
@romintomasetti Let me discuss this change with @CamelliaDPG  first. 
A minor comment: In Intrepid2, the first parameters of a function are typically output parameters, so I would not pass the execution space as the first parameter.

**Comment by romintomasetti** [comment]
> @romintomasetti Let me discuss this change with @CamelliaDPG first.

Sure, no problem :smile:

> A minor comment: In Intrepid2, the first parameters of a function are typically output parameters, so I would not pass the execution space as the first parameter.

Indeed. However, I think most of `Kokkos`/`Kokkos Kernels` routines that take the `space` argument put it first whatsoever. Take for instance `packages/kokkos-kernels/sparse/src/KokkosSparse_SortCrs.hpp`.

I think it would be nice to align on this :smile:

**Comment by mperego** [comment]
@romintomasetti Nate and I talked and we are fine with the changes that you propose. Thanks for helping us developing Intrepid2. We recommend adding unit tests that exercise this new feature.

**Comment by romintomasetti** [comment]
Hi @mperego !

I think it will take too much of my time to change all the `Intrepid2` bases.

Here is my proposal: I will only change those that we use in our group, and the relevant tests in `packages/intrepid2/unit-test/Discretization/Basis`:
- [ ] `HCURL TRI In`
- [ ] `HGRAD LINE C1 C2 Cn`
- [ ] `HGRAD QUAD C1 C2 Cn`
- [ ] `HGRAD TRI C1 C2 Cn`

and I would make small PRs later on if I need other bases. Would that be OK for you ?

**Comment by mperego** [comment]
> Hi @mperego !
> 
> I think it will take too much of my time to change all the `Intrepid2` bases.
> 
> Here is my proposal: I will only change those that we use in our group, and the relevant tests in `packages/intrepid2/unit-test/Discretization/Basis`:
> 
>     * [ ]  `HCURL TRI In`
> 
>     * [ ]  `HGRAD LINE C1 C2 Cn`
> 
>     * [ ]  `HGRAD QUAD C1 C2 Cn`
> 
>     * [ ]  `HGRAD TRI C1 C2 Cn`
> 
> 
> and I would make small PRs later on if I need other bases. Would that be OK for you ?

This is fine. The test should exercise the case where the execution space passed to `getValues` is different than the default `ExecutionSpace{}`.

**Comment by romintomasetti** [comment]
> This is fine. The test should exercise the case where the execution space passed to `getValues` is different than the default `ExecutionSpace{}`.

Ok, I create each time a new space instance using `Kokkos::Experimental::partition_space`. That should be enough to get a non-default instance (*e.g* a new stream for `Cuda`).

**Comment by mperego** [comment]
@romintomasetti let me know when you are done with your changes and ready for the code to be tested.

File: packages/intrepid2/unit-test/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,5 @@
+TRIBITS_INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR})
```
Comment (romintomasetti) [inline packages/intrepid2/unit-test/CMakeLists.txt:1]: used to allow
```c++
#include "packages/intrepid2/unit-test/Discretization/Basis/Macros.hpp"
```
in the test files, but I'm not sure it is "standard" in Trilinos to do that.

**Comment by romintomasetti** [comment]
> @romintomasetti let me know when you are done with your changes and ready for the code to be tested.

@mperego I think it's ready for a first round :)

**Comment by mperego** [comment]
@romintomasetti Here is the output for one of the errors:
```
<b>Test: </b> <a id="summary_link" href="https://trilinos-cdash.sandia.gov/testSummary.php?project=1&amp;name=Intrepid2_unit-test_Discretization_Basis_HGRAD_TET_C1_FEM_test_01_Serial_DOUBLE_MPI_1&amp;date=2023-10-11">
      Intrepid2_unit-test_Discretization_Basis_HGRAD_TET_C1_FEM_test_01_Serial_DOUBLE_MPI_1
    </a> <b class="error-text">
      (Failed)
    </b> <br> <b>Build: </b> <a id="build_link" href="https://trilinos-cdash.sandia.gov/build/1191489">
      
PR-12366-test-rhel7_sems-cuda-11.4.2-sems-gnu-10.1.0-sems-openmpi-4.0.5_release_static_Volta70_no-asan_complex_no-fpic_mpi_pt_no-rdc_no-uvm_deprecated-on_no-package-enables-2570
    </a> <a id="site_link" href="https://trilinos-cdash.sandia.gov/viewSite.php?siteid=99">
      (ascicgpu036)
    </a>
    on 2023-10-11 09:12:35
    <br> <div><b>Repository revision: </b> <a id="revision_link" href="https://github.com/trilinos/Trilinos//commit/f915f9b2e99ffe2c257683b95895db657b7a36e3">
        f915f9b2e99ffe2c257683b95895db657b7a36e3
      </a> <br></div> <div><b>Test Details: </b>
      Completed (Failed)
      <br></div> <div><br> <b>Test Timing: </b> <b class="normal-text">
        Passed
      </b> </div> <div><b>Labels: </b>
      Intrepid2
      <br></div> <br> 

Processors | 1
-- | --
24
Required regular expression not found. Regex=[TEST PASSED<br /> ]

 <br> <img src="https://trilinos-cdash.sandia.gov/img/console.png"> <a id="commandlinelink" href="https://trilinos-cdash.sandia.gov/test/37769497#"><span>Show Command Line</span> </a>  <br> <div><img src="https://trilinos-cdash.sandia.gov/img/console.png"> <a id="environmentlink" href="https://trilinos-cdash.sandia.gov/test/37769497#"><span>Show Environment</span> </a>  <br></div> <img src="https://trilinos-cdash.sandia.gov/img/graph.png">
    Display graphs:
    <br>    <br> <b>Test output</b> <pre id="test_output">Warning: environment variable 'KOKKOS_NUM_DEVICES' is deprecated. Use 'KOKKOS_MAP_DEVICE_ID_BY=mpi_rank' instead. Raised by Kokkos::initialize().
DeviceSpace::  Host Serial Execution Space:
  KOKKOS_ENABLE_SERIAL: yes

Serial Runtime Configuration:
HostSpace::    Host Serial Execution Space:
  KOKKOS_ENABLE_SERIAL: yes

Serial Runtime Configuration:
================================================================================
|                                                                              |
|                 Unit Test (Basis_HGRAD_TET_C1_FEM)                           |
|                                                                              |
|     1) Conversion of Dof tags into Dof ordinals and back                     |
|     2) Basis values for VALUE, GRAD, CURL, and Dk operators                  |
|                                                                              |
|  Questions? Contact  Pavel Bochev  (pbboche@sandia.gov),                     |
|                      Robert Kirby  (robert.c.kirby@ttu.edu),                 |
|                      Denis Ridzal  (dridzal@sandia.gov),                     |
|                      Kara Peterson (dridzal@sandia.gov),                     |
|                      Mauro Perego  (mperego@sandia.gov),                     |
|                      Kyungjoo Kim  (kyukim@sandia.gov).                      |
|                                                                              |
|  Intrepid's website: http://trilinos.sandia.gov/packages/intrepid            |
|  Trilinos website:   http://trilinos.sandia.gov                              |
|                                                                              |
================================================================================

===============================================================================
| TEST 1: constructors and exceptions                                         |
===============================================================================

===============================================================================
| TEST 2: correctness of tag to enum and enum to tag lookups                  |
===============================================================================

===============================================================================
| TEST 3: correctness of basis function values                                |
===============================================================================
[Intrepid2] Error in file ../Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp, line 440
            Test that evaluated to true: true
            &gt;&gt;&gt; ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes. 
&gt;&gt;&gt; ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes.


===============================================================================
| TEST 4: correctness of DoF locations                                        |
===============================================================================
[Intrepid2] Error in file ../Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp, line 440
            Test that evaluated to true: true
            &gt;&gt;&gt; ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes. 
&gt;&gt;&gt; ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes.


===============================================================================
| TEST 5: Function Space is Correct                                           |
===============================================================================
End Result: TEST FAILED
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[41385,1],0]
  Exit code:    24
--------------------------------------------------------------------------
</pre>Test: Intrepid2_unit-test_Discretization_Basis_HGRAD_TET_C1_FEM_test_01_Serial_DOUBLE_MPI_1 (Failed)
Build: PR-12366-test-rhel7_sems-cuda-11.4.2-sems-gnu-10.1.0-sems-openmpi-4.0.5_release_static_Volta70_no-asan_complex_no-fpic_mpi_pt_no-rdc_no-uvm_deprecated-on_no-package-enables-2570[ (ascicgpu036) ](https://trilinos-cdash.sandia.gov/viewSite.php?siteid=99)on 2023-10-11 09:12:35
Repository revision: f915f9b2e99ffe2c257683b95895db657b7a36e3
Test Details: Completed (Failed)

Test Timing: Passed
Labels: Intrepid2

Processors 	1
Exit Value 	24
Fail Reason 	Required regular expression not found. Regex=[TEST PASSED<br /> ]

Show Command Line
Show Environment
Display graphs:

Test output

Warning: environment variable 'KOKKOS_NUM_DEVICES' is deprecated. Use 'KOKKOS_MAP_DEVICE_ID_BY=mpi_rank' instead. Raised by Kokkos::initialize().
DeviceSpace::  Host Serial Execution Space:
  KOKKOS_ENABLE_SERIAL: yes

Serial Runtime Configuration:
HostSpace::    Host Serial Execution Space:
  KOKKOS_ENABLE_SERIAL: yes

Serial Runtime Configuration:
================================================================================
|                                                                              |
|                 Unit Test (Basis_HGRAD_TET_C1_FEM)                           |
|                                                                              |
|     1) Conversion of Dof tags into Dof ordinals and back                     |
|     2) Basis values for VALUE, GRAD, CURL, and Dk operators                  |
|                                                                              |
|  Questions? Contact  Pavel Bochev  (pbboche@sandia.gov),                     |
|                      Robert Kirby  (robert.c.kirby@ttu.edu),                 |
|                      Denis Ridzal  (dridzal@sandia.gov),                     |
|                      Kara Peterson (dridzal@sandia.gov),                     |
|                      Mauro Perego  (mperego@sandia.gov),                     |
|                      Kyungjoo Kim  (kyukim@sandia.gov).                      |
|                                                                              |
|  Intrepid's website: http://trilinos.sandia.gov/packages/intrepid            |
|  Trilinos website:   http://trilinos.sandia.gov                              |
|                                                                              |
================================================================================

===============================================================================
| TEST 1: constructors and exceptions                                         |
===============================================================================

===============================================================================
| TEST 2: correctness of tag to enum and enum to tag lookups                  |
===============================================================================

===============================================================================
| TEST 3: correctness of basis function values                                |
===============================================================================
[Intrepid2] Error in file ../Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp, line 440
            Test that evaluated to true: true
            >>> ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes. 
>>> ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes.


===============================================================================
| TEST 4: correctness of DoF locations                                        |
===============================================================================
[Intrepid2] Error in file ../Trilinos/packages/intrepid2/src/Discretization/Basis/Intrepid2_Basis.hpp, line 440
            Test that evaluated to true: true
            >>> ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes. 
>>> ERROR (Basis::getValues): this method (FEM) is not supported or should be overridden accordingly by derived classes.


===============================================================================
| TEST 5: Function Space is Correct                                           |
===============================================================================
End Result: TEST FAILED
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[41385,1],0]
  Exit code:    24
--------------------------------------------------------------------------
```

There are other errors but they might be temporary issues

**Comment by romintomasetti** [comment]
Ok, indeed. I think it should be OK now :smile:  (last time I compiled everything but seems I forgot to run some of the tests)

**Comment by mperego** [comment]
@trilinos/framework Can you help us understand why the `TrilinosInstallTests_reduced_tarball ` test failed?

**Comment by sebrowne** [comment]
@mperego the issue is that we set `Trilinos_ENABLE_Intrepid=OFF` in this build to disable deprecated packages (Intrepid is going to be deprecated next year).  However, the PR system managed to re-enable it in the main build phase because of changed code.  Then, once it got to the reduced tarball test, it was again disabled because there was nothing explicitly re-enabling it, and the test caught that some of the packages were not enabled.  The issue here is how `Trilinos_ENABLE_Intrepid=ON` is getting set instead of OFF.  It has to do with CMake cachefiles and variable scope precedence I believe.  I will see if I can figure it out and report back.

**Comment by mperego** [comment]
@sebrowne, OK I see. The PR has changes to Intrepid (text formatting) that are not needed. 
@romintomasetti would you mind revert the changes in Intrepid? This should fix the issue with the tarball.

**Comment by sebrowne** [comment]
@mperego if you want to do that here, that should fix this particular PR.  I will try to fix this going forwards by forcing Intrepid, etc. to be disabled here even if they get changed (that's the intent after all).  I think I know how to do that, just need to test it locally.

**Comment by romintomasetti** [comment]
> @sebrowne, OK I see, thanks. The PR has changes to Intrepid (text formatting) that are not needed. @romintomasetti would you mind reverting the changes in Intrepid? This should fix the issue with the tarball.

Oups sorry for that interrelated change! I reverted the changes in `Intrepid`.

**Comment by romintomasetti** [comment]
#12395 apparently introduced conflicts. Let me solve them.

**Review by CamelliaDPG — APPROVED** [review]
Looks good to me.  Thanks, @romintomasetti!

**Pull request merged**